File: LC50/lc01.py
"""
Given an array of integers,
 return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution,
 and you may not use the same element twice.

Example:
Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
"""


def two_sum(nums, target):
    note = {}
    for i, number in enumerate(nums):
        if number in note:
            j = note[number]
            if i < j:
                return [i, j]
            return [j, i]
        else:
            counter = target - number
            note[counter] = i


def test():
    nums_list = [
        [3, 2, 4],
        [2, 11, 53, 7],
        [1, 3, 4, 5, 42, 16, 17, 12]
    ]
    target_list = [
        6, 9, 16
    ]

    for nums, target in zip(nums_list, target_list):
        print(two_sum(nums, target))


test()

File: LC50/lc01.js
function two_sum(nums, target) {
    var note = {};
    for (var i=0; i<nums.length; i++) {
        var number = nums[i];
        if (note.hasOwnProperty(number)) {
            var j = note[number];
            if (i < j) {
                return [i, j];
            }
            return [j, i];
        } else {
            note[target - number] = i;
        }
    } 
}


function test() {
    var nums_list = [
        [3, 2, 4],
        [2, 11, 53, 7],
        [1, 3, 4, 5, 42, 16, 17, 12]
    ];
    var target_list = [
        6, 9, 16
    ];
    for (var i=0; i<3; i++) {
        var nums = nums_list[i];
        var target = target_list[i];
        console.log(two_sum(nums, target));
    }
}


test();